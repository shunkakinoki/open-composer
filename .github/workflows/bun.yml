name: Bun
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
  workflow_dispatch:
concurrency:
  group: ${{ github.workflow }}-${{ github.event_name }}-${{ github.head_ref || github.sha }}
  cancel-in-progress: true
jobs:
  bun-build:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Build
        run: bun run build
  bun-lint:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Check
        run: bun run check
      - name: Format
        run: bun run format
      - name: Lint
        run: bun run lint
      - name: Type Check
        run: bun run type-check
  bun-test:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Test
        run: bun run test
  bun-check:
    if: always()
    needs:
      - bun-build
      - bun-lint
      - bun-test
    runs-on: ubuntu-latest
    timeout-minutes: 3
    steps:
      - name: Alls Green
        uses: re-actors/alls-green@release/v1
        with:
          jobs: ${{ toJSON(needs) }}
