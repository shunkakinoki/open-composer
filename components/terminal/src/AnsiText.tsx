import React from 'react';
import { Box, Text } from 'ink';
import type { AnsiOutput, AnsiToken } from './terminalSerializer.js';

/**
 * Props for AnsiText component
 */
export interface AnsiTextProps {
  /** Serialized terminal output */
  output: AnsiOutput;
  /** Maximum width for wrapping */
  width?: number;
  /** Maximum height (number of lines to show) */
  height?: number;
}

/**
 * Renders a single styled token
 */
const StyledToken: React.FC<{ token: AnsiToken }> = ({ token }) => {
  const { text, bold, italic, underline, dim, inverse, fg, bg } = token;

  // Build style props for Ink Text component
  const props: any = {};

  if (bold) props.bold = true;
  if (italic) props.italic = true;
  if (underline) props.underline = true;
  if (dim) props.dimColor = true;
  if (inverse) props.inverse = true;

  // Handle colors - Ink supports both named colors and hex
  if (fg && fg !== '') {
    props.color = fg;
  }
  if (bg && bg !== '') {
    props.backgroundColor = bg;
  }

  return <Text {...props}>{text}</Text>;
};

/**
 * Renders serialized terminal output with ANSI styling in Ink
 *
 * This component takes AnsiOutput (from terminalSerializer) and renders
 * it in Ink with proper text styling, colors, and formatting.
 *
 * @example
 * ```tsx
 * <AnsiText output={serializedOutput} width={80} height={24} />
 * ```
 */
export const AnsiText: React.FC<AnsiTextProps> = ({
  output,
  width,
  height
}) => {
  // Apply height limit if specified
  const displayLines = height ? output.slice(0, height) : output;

  return (
    <Box flexDirection="column" width={width}>
      {displayLines.map((line, lineIndex) => (
        <Box key={lineIndex}>
          {line.map((token, tokenIndex) => (
            <StyledToken key={tokenIndex} token={token} />
          ))}
        </Box>
      ))}
    </Box>
  );
};
