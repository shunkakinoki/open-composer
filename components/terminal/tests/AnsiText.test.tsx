import { describe, it, expect } from 'bun:test';
import { render } from 'ink-testing-library';
import React from 'react';
import { AnsiText } from '../src/AnsiText.js';
import type { AnsiOutput } from '../src/terminalSerializer.js';

describe('AnsiText', () => {
  describe('snapshot tests', () => {
    it('should render empty output', () => {
      const output: AnsiOutput = [];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render plain text', () => {
      const output: AnsiOutput = [
        [{ text: 'Hello, World!' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      const frame = lastFrame();
      expect(frame).toContain('Hello, World!');
      expect(frame).toMatchSnapshot();
    });

    it('should render multiple lines', () => {
      const output: AnsiOutput = [
        [{ text: 'Line 1' }],
        [{ text: 'Line 2' }],
        [{ text: 'Line 3' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      const frame = lastFrame();
      expect(frame).toContain('Line 1');
      expect(frame).toContain('Line 2');
      expect(frame).toContain('Line 3');
      expect(frame).toMatchSnapshot();
    });

    it('should render text with bold styling', () => {
      const output: AnsiOutput = [
        [{ text: 'Bold Text', bold: true }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render text with italic styling', () => {
      const output: AnsiOutput = [
        [{ text: 'Italic Text', italic: true }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render text with underline styling', () => {
      const output: AnsiOutput = [
        [{ text: 'Underlined Text', underline: true }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render text with dim styling', () => {
      const output: AnsiOutput = [
        [{ text: 'Dim Text', dim: true }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render text with inverse styling', () => {
      const output: AnsiOutput = [
        [{ text: 'Inverse Text', inverse: true }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render text with foreground color', () => {
      const output: AnsiOutput = [
        [{ text: 'Red Text', fg: '#ff0000' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render text with background color', () => {
      const output: AnsiOutput = [
        [{ text: 'Blue Background', bg: '#0000ff' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render text with multiple styles', () => {
      const output: AnsiOutput = [
        [{
          text: 'Styled Text',
          bold: true,
          italic: true,
          underline: true,
          fg: '#00ff00',
          bg: '#000000'
        }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render mixed styled tokens on same line', () => {
      const output: AnsiOutput = [
        [
          { text: 'Normal ' },
          { text: 'Bold', bold: true },
          { text: ' ' },
          { text: 'Red', fg: '#ff0000' },
        ],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      const frame = lastFrame();
      expect(frame).toContain('Normal');
      expect(frame).toContain('Bold');
      expect(frame).toContain('Red');
      expect(frame).toMatchSnapshot();
    });

    it('should render with width constraint', () => {
      const output: AnsiOutput = [
        [{ text: 'This is a very long line that should be constrained by width' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} width={40} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('should render with height constraint', () => {
      const output: AnsiOutput = [
        [{ text: 'Line 1' }],
        [{ text: 'Line 2' }],
        [{ text: 'Line 3' }],
        [{ text: 'Line 4' }],
        [{ text: 'Line 5' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} height={3} />);

      const frame = lastFrame();
      expect(frame).toContain('Line 1');
      expect(frame).toContain('Line 2');
      expect(frame).toContain('Line 3');
      expect(frame).not.toContain('Line 4');
      expect(frame).not.toContain('Line 5');
      expect(frame).toMatchSnapshot();
    });

    it('should render complex output with multiple lines and tokens', () => {
      const output: AnsiOutput = [
        [{ text: '$ ', fg: '#00ff00', bold: true }, { text: 'echo "Hello"' }],
        [{ text: 'Hello' }],
        [{ text: '$ ', fg: '#00ff00', bold: true }, { text: 'ls -la' }],
        [{ text: 'total 42', bold: true }],
        [{ text: 'drwxr-xr-x', fg: '#0000ff' }, { text: '  5 user  group  160 Jan  1 12:00 ' }, { text: '.', fg: '#0000ff', bold: true }],
      ];
      const { lastFrame } = render(<AnsiText output={output} width={80} />);

      expect(lastFrame()).toMatchSnapshot();
    });
  });

  describe('integration tests', () => {
    it('should handle empty text tokens', () => {
      const output: AnsiOutput = [
        [{ text: '' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toBeDefined();
    });

    it('should handle lines with no tokens', () => {
      const output: AnsiOutput = [
        [],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toBeDefined();
    });

    it('should handle undefined colors gracefully', () => {
      const output: AnsiOutput = [
        [{ text: 'Text', fg: undefined, bg: undefined }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toContain('Text');
    });

    it('should handle empty string colors gracefully', () => {
      const output: AnsiOutput = [
        [{ text: 'Text', fg: '', bg: '' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toContain('Text');
    });

    it('should respect both width and height constraints together', () => {
      const output: AnsiOutput = [
        [{ text: 'Line 1 with some text' }],
        [{ text: 'Line 2 with some text' }],
        [{ text: 'Line 3 with some text' }],
        [{ text: 'Line 4 with some text' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} width={20} height={2} />);

      const frame = lastFrame();
      expect(frame).toContain('Line 1');
      expect(frame).toContain('Line 2');
      expect(frame).not.toContain('Line 3');
      expect(frame).not.toContain('Line 4');
    });

    it('should update when output changes', () => {
      const initialOutput: AnsiOutput = [
        [{ text: 'Initial' }],
      ];

      const { lastFrame, rerender } = render(<AnsiText output={initialOutput} />);

      expect(lastFrame()).toContain('Initial');

      const updatedOutput: AnsiOutput = [
        [{ text: 'Updated' }],
      ];

      rerender(<AnsiText output={updatedOutput} />);

      expect(lastFrame()).toContain('Updated');
      expect(lastFrame()).not.toContain('Initial');
    });

    it('should update when width changes', () => {
      const output: AnsiOutput = [
        [{ text: 'Some text content that wraps' }],
      ];

      const { rerender } = render(<AnsiText output={output} width={80} />);

      // Should not throw when width changes
      rerender(<AnsiText output={output} width={40} />);
    });

    it('should update when height changes', () => {
      const output: AnsiOutput = [
        [{ text: 'Line 1' }],
        [{ text: 'Line 2' }],
        [{ text: 'Line 3' }],
        [{ text: 'Line 4' }],
      ];

      const { lastFrame, rerender } = render(<AnsiText output={output} height={4} />);

      expect(lastFrame()).toContain('Line 4');

      rerender(<AnsiText output={output} height={2} />);

      expect(lastFrame()).not.toContain('Line 4');
    });
  });

  describe('edge cases', () => {
    it('should handle very long lines', () => {
      const longText = 'a'.repeat(1000);
      const output: AnsiOutput = [
        [{ text: longText }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toBeDefined();
    });

    it('should handle many lines', () => {
      const output: AnsiOutput = Array.from({ length: 100 }, (_, i) => [
        { text: `Line ${i + 1}` },
      ]);
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toBeDefined();
    });

    it('should handle many tokens on a single line', () => {
      const output: AnsiOutput = [
        Array.from({ length: 100 }, (_, i) => ({
          text: `Token${i} `,
          fg: i % 2 === 0 ? '#ff0000' : '#00ff00',
        })),
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toBeDefined();
    });

    it('should handle special characters', () => {
      const output: AnsiOutput = [
        [{ text: 'Tab:\t' }],
        [{ text: 'Newline:\n' }],
        [{ text: 'Special: ðŸš€ âœ¨ ðŸŽ‰' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} />);

      expect(lastFrame()).toBeDefined();
    });

    it('should handle zero width', () => {
      const output: AnsiOutput = [
        [{ text: 'Text' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} width={0} />);

      expect(lastFrame()).toBeDefined();
    });

    it('should handle zero height', () => {
      const output: AnsiOutput = [
        [{ text: 'Text' }],
      ];
      const { lastFrame } = render(<AnsiText output={output} height={0} />);

      // With height 0 (falsy), the component defaults to showing all output
      // This is expected behavior since 0 is falsy in the ternary check
      const frame = lastFrame();
      expect(frame).toBeDefined();
      expect(frame).toContain('Text');
    });
  });
});
