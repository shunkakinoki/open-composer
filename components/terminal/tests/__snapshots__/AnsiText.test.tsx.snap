// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`AnsiText snapshot tests should render empty output 1`] = `""`;

exports[`AnsiText snapshot tests should render plain text 1`] = `"Hello, World!"`;

exports[`AnsiText snapshot tests should render multiple lines 1`] = `
"Line 1
Line 2
Line 3"
`;

exports[`AnsiText snapshot tests should render text with bold styling 1`] = `"Bold Text"`;

exports[`AnsiText snapshot tests should render text with italic styling 1`] = `"Italic Text"`;

exports[`AnsiText snapshot tests should render text with underline styling 1`] = `"Underlined Text"`;

exports[`AnsiText snapshot tests should render text with dim styling 1`] = `"Dim Text"`;

exports[`AnsiText snapshot tests should render text with inverse styling 1`] = `"Inverse Text"`;

exports[`AnsiText snapshot tests should render text with foreground color 1`] = `"Red Text"`;

exports[`AnsiText snapshot tests should render text with background color 1`] = `"Blue Background"`;

exports[`AnsiText snapshot tests should render text with multiple styles 1`] = `"Styled Text"`;

exports[`AnsiText snapshot tests should render mixed styled tokens on same line 1`] = `"Normal Bold Red"`;

exports[`AnsiText snapshot tests should render with width constraint 1`] = `
"This is a very long line that should be
constrained by width"
`;

exports[`AnsiText snapshot tests should render with height constraint 1`] = `
"Line 1
Line 2
Line 3"
`;

exports[`AnsiText snapshot tests should render complex output with multiple lines and tokens 1`] = `
"$ echo "Hello"
Hello
$ ls -la
total 42
drwxr-xr-x  5 user  group  160 Jan  1 12:00 ."
`;
